File: example/src/main/java/com/example/collapsingtoolbarlayoutsubtitle/ArticleActivity.java
package com.example.collapsingtoolbarlayoutsubtitle;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.SubtitleCollapsingToolbarLayout;
import android.support.v7.widget.Toolbar;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.afollestad.materialdialogs.MaterialDialog;
import com.hendraanggrian.bundler.annotations.BindExtra;

import java.util.LinkedHashMap;
import java.util.Map;

import butterknife.BindView;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public class ArticleActivity extends BaseActivity implements View.OnClickListener {

    private static final String OPTION_SET_TITLE = "Set title";
    private static final String OPTION_SET_SUBTITLE = "Set subtitle";
    private static final String OPTION_SET_EXPANDED_GRAVITY = "Set expanded gravity";
    private static final String OPTION_SET_COLLAPSED_GRAVITY = "Set collapsed gravity";
    private static final String OPTION_DISABLE_BACK_BUTTON = "Disable back button";
    private static final Map<CharSequence, Integer> GRAVITY = new LinkedHashMap<>();

    static {
        GRAVITY.put("START", Gravity.START);
        GRAVITY.put("TOP", Gravity.TOP);
        GRAVITY.put("END", Gravity.END);
        GRAVITY.put("BOTTOM", Gravity.BOTTOM);
        GRAVITY.put("CENTER_HORIZONTAL", Gravity.CENTER_HORIZONTAL);
        GRAVITY.put("CENTER_VERTICAL", Gravity.CENTER_VERTICAL);
        GRAVITY.put("CENTER", Gravity.CENTER);
    }

    @BindExtra int layoutRes;
    @BindView(R.id.collapsingtoolbarlayout_article) View collapsingToolbarLayout;
    @BindView(R.id.toolbar_article) Toolbar toolbar;
    @BindView(R.id.floatingactionbutton_article) FloatingActionButton floatingActionButton;

    @Override
    int getContentView() {
        return layoutRes;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setSupportActionBar(toolbar);
        floatingActionButton.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        final BaseToolbarLayout toolbarLayout = getToolbarLayout();
        new MaterialDialog.Builder(this)
                .items(OPTION_SET_TITLE, OPTION_SET_SUBTITLE, OPTION_SET_EXPANDED_GRAVITY, OPTION_SET_COLLAPSED_GRAVITY, OPTION_DISABLE_BACK_BUTTON)
                .itemsCallback(new MaterialDialog.ListCallback() {
                    @Override
                    public void onSelection(MaterialDialog dialog, View itemView, int position, final CharSequence text) {
                        switch (text.toString()) {
                            case OPTION_SET_TITLE:
                            case OPTION_SET_SUBTITLE:
                                new MaterialDialog.Builder(ArticleActivity.this)
                                        .input("Text", "", new MaterialDialog.InputCallback() {
                                            @Override
                                            public void onInput(@NonNull MaterialDialog dialog, CharSequence input) {
                                                switch (text.toString()) {
                                                    case OPTION_SET_TITLE:
                                                        toolbarLayout.setTitle(input);
                                                        break;
                                                    case OPTION_SET_SUBTITLE:
                                                        if (collapsingToolbarLayout instanceof SubtitleCollapsingToolbarLayout) {
                                                            ((SubtitleCollapsingToolbarLayout) collapsingToolbarLayout).setSubtitle(input);
                                                            return;
                                                        }
                                                        Toast.makeText(ArticleActivity.this, "Unsupported", Toast.LENGTH_SHORT).show();
                                                        break;
                                                }
                                            }
                                        })
                                        .show();
                                break;
                            case OPTION_SET_EXPANDED_GRAVITY:
                            case OPTION_SET_COLLAPSED_GRAVITY:
                                new MaterialDialog.Builder(ArticleActivity.this)
                                        .items(GRAVITY.keySet())
                                        .itemsCallbackMultiChoice(null, new MaterialDialog.ListCallbackMultiChoice() {
                                            @Override
                                            public boolean onSelection(MaterialDialog dialog, Integer[] which, CharSequence[] texts) {
                                                Integer flags = null;
                                                for (CharSequence text : texts) {
                                                    int flag = GRAVITY.get(text);
                                                    if (flags == null)
                                                        flags = flag;
                                                    else
                                                        flags |= flag;
                                                }
                                                if (flags != null) {
                                                    switch (text.toString()) {
                                                        case OPTION_SET_EXPANDED_GRAVITY:
                                                            toolbarLayout.setExpandedTitleGravity(flags);
                                                            break;
                                                        case OPTION_SET_COLLAPSED_GRAVITY:
                                                            toolbarLayout.setCollapsedTitleGravity(flags);
                                                            break;
                                                    }
                                                }
                                                return false;
                                            }
                                        })
                                        .positiveText(android.R.string.ok)
                                        .show();
                                break;
                            case OPTION_DISABLE_BACK_BUTTON:
                                toolbar.setNavigationIcon(null);
                                break;
                        }
                    }
                })
                .show();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home)
            finish();
        return super.onOptionsItemSelected(item);
    }

    @NonNull
    BaseToolbarLayout getToolbarLayout() {
        return new BaseToolbarLayout() {
            @Override
            public void setTitle(CharSequence text) {
                if (collapsingToolbarLayout instanceof CollapsingToolbarLayout)
                    ((CollapsingToolbarLayout) collapsingToolbarLayout).setTitle(text);
                else if (collapsingToolbarLayout instanceof SubtitleCollapsingToolbarLayout)
                    ((SubtitleCollapsingToolbarLayout) collapsingToolbarLayout).setTitle(text);
            }

            @Override
            public void setExpandedTitleGravity(int gravity) {
                if (collapsingToolbarLayout instanceof CollapsingToolbarLayout)
                    ((CollapsingToolbarLayout) collapsingToolbarLayout).setExpandedTitleGravity(gravity);
                else if (collapsingToolbarLayout instanceof SubtitleCollapsingToolbarLayout)
                    ((SubtitleCollapsingToolbarLayout) collapsingToolbarLayout).setExpandedTitleGravity(gravity);
            }

            @Override
            public void setCollapsedTitleGravity(int gravity) {
                if (collapsingToolbarLayout instanceof CollapsingToolbarLayout)
                    ((CollapsingToolbarLayout) collapsingToolbarLayout).setCollapsedTitleGravity(gravity);
                else if (collapsingToolbarLayout instanceof SubtitleCollapsingToolbarLayout)
                    ((SubtitleCollapsingToolbarLayout) collapsingToolbarLayout).setCollapsedTitleGravity(gravity);
            }
        };
    }
}

File: example/src/main/java/com/example/collapsingtoolbarlayoutsubtitle/ArticleActivity.kt
package com.example.collapsingtoolbarlayoutsubtitle

import android.os.Bundle
import android.support.design.widget.CollapsingToolbarLayout
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.SubtitleCollapsingToolbarLayout
import android.support.v7.widget.Toolbar
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import butterknife.BindView
import com.afollestad.materialdialogs.MaterialDialog
import com.hendraanggrian.bundler.BindExtra
import com.hendraanggrian.support.utils.widget.Toasts
import java.util.*

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
class ArticleActivity : BaseActivity(), View.OnClickListener {

    override val contentView: Int
        get() = layoutRes

    @JvmField @BindExtra var layoutRes: Int = 0
    @BindView(R.id.collapsingtoolbarlayout_article) lateinit var collapsingToolbarLayout: View
    @BindView(R.id.toolbar_article) lateinit var toolbar: Toolbar
    @BindView(R.id.floatingactionbutton_article) lateinit var floatingActionButton: FloatingActionButton
    var menuItem: MenuItem? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setSupportActionBar(toolbar)
        floatingActionButton.setOnClickListener(this)

        if (collapsingToolbarLayout is SubtitleCollapsingToolbarLayout) {
            (collapsingToolbarLayout as SubtitleCollapsingToolbarLayout).setCollapsedTitleColorAttr(R.attr.colorAccent)
        }
    }

    override fun onClick(v: View) {
        val toolbarLayout = toolbarLayout
        MaterialDialog.Builder(this)
                .items(OPTION_SET_TITLE,
                        OPTION_SET_SUBTITLE,
                        OPTION_SET_EXPANDED_GRAVITY,
                        OPTION_SET_COLLAPSED_GRAVITY,
                        OPTION_DISABLE_BACK_BUTTON,
                        OPTION_TOGGLE_MENU_ITEM)
                .itemsCallback { _, _, _, text ->
                    when (text.toString()) {
                        OPTION_SET_TITLE, OPTION_SET_SUBTITLE -> MaterialDialog.Builder(this@ArticleActivity)
                                .input("Text", "") { _, input ->
                                    when (text.toString()) {
                                        OPTION_SET_TITLE -> toolbarLayout.setTitle(input)
                                        OPTION_SET_SUBTITLE -> toolbarLayout.setSubtitle(input)
                                    }
                                }
                                .show()
                        OPTION_SET_EXPANDED_GRAVITY, OPTION_SET_COLLAPSED_GRAVITY -> MaterialDialog.Builder(this@ArticleActivity)
                                .items(GRAVITY.keys)
                                .itemsCallbackMultiChoice(null) { dialog, which, texts ->
                                    var flags: Int? = null
                                    for (text in texts) {
                                        val flag = GRAVITY[text]
                                        if (flags == null)
                                            flags = flag
                                        else
                                            flags = flags or flag as Int
                                    }
                                    if (flags != null) {
                                        when (text.toString()) {
                                            OPTION_SET_EXPANDED_GRAVITY -> toolbarLayout.setExpandedTitleGravity(flags)
                                            OPTION_SET_COLLAPSED_GRAVITY -> toolbarLayout.setCollapsedTitleGravity(flags)
                                        }
                                    }
                                    false
                                }
                                .positiveText(android.R.string.ok)
                                .show()
                        OPTION_DISABLE_BACK_BUTTON -> toolbar.navigationIcon = null
                        OPTION_TOGGLE_MENU_ITEM -> menuItem!!.isVisible = !menuItem!!.isVisible
                    }
                }
                .show()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.article, menu)
        menuItem = menu.findItem(R.id.item_article_bookmark)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home)
            finish()
        return super.onOptionsItemSelected(item)
    }

    internal val toolbarLayout: BaseToolbarLayout
        get() = object : BaseToolbarLayout {
            override fun setTitle(text: CharSequence?) {
                if (collapsingToolbarLayout is CollapsingToolbarLayout)
                    (collapsingToolbarLayout as CollapsingToolbarLayout).title = text
                else if (collapsingToolbarLayout is SubtitleCollapsingToolbarLayout)
                    (collapsingToolbarLayout as SubtitleCollapsingToolbarLayout).title = text
            }

            override fun setSubtitle(text: CharSequence?) {
                if (collapsingToolbarLayout is CollapsingToolbarLayout)
                    Toasts.showShort(this@ArticleActivity, "Unsupported.")
                else if (collapsingToolbarLayout is SubtitleCollapsingToolbarLayout)
                    (collapsingToolbarLayout as SubtitleCollapsingToolbarLayout).subtitle = text
            }

            override fun setExpandedTitleGravity(gravity: Int) {
                if (collapsingToolbarLayout is CollapsingToolbarLayout)
                    (collapsingToolbarLayout as CollapsingToolbarLayout).expandedTitleGravity = gravity
                else if (collapsingToolbarLayout is SubtitleCollapsingToolbarLayout)
                    (collapsingToolbarLayout as SubtitleCollapsingToolbarLayout).expandedTitleGravity = gravity
            }

            override fun setCollapsedTitleGravity(gravity: Int) {
                if (collapsingToolbarLayout is CollapsingToolbarLayout)
                    (collapsingToolbarLayout as CollapsingToolbarLayout).collapsedTitleGravity = gravity
                else if (collapsingToolbarLayout is SubtitleCollapsingToolbarLayout)
                    (collapsingToolbarLayout as SubtitleCollapsingToolbarLayout).collapsedTitleGravity = gravity
            }
        }

    companion object {
        private val OPTION_SET_TITLE = "Set title"
        private val OPTION_SET_SUBTITLE = "Set subtitle"
        private val OPTION_SET_EXPANDED_GRAVITY = "Set expanded gravity"
        private val OPTION_SET_COLLAPSED_GRAVITY = "Set collapsed gravity"
        private val OPTION_DISABLE_BACK_BUTTON = "Disable back button"
        private val OPTION_TOGGLE_MENU_ITEM = "Toggle menu item visible"
        private val GRAVITY = LinkedHashMap<CharSequence, Int>()

        init {
            GRAVITY.put("START", Gravity.START)
            GRAVITY.put("TOP", Gravity.TOP)
            GRAVITY.put("END", Gravity.END)
            GRAVITY.put("BOTTOM", Gravity.BOTTOM)
            GRAVITY.put("CENTER_HORIZONTAL", Gravity.CENTER_HORIZONTAL)
            GRAVITY.put("CENTER_VERTICAL", Gravity.CENTER_VERTICAL)
            GRAVITY.put("CENTER", Gravity.CENTER)
        }
    }
}